### PR TITLE
ADD: New function for calculating PBL using the Heffter method

### PR DESCRIPTION
**Pull request by AdamTheisen**
This includes a second method (Heffter) for calculating PBL from radiosonde data.  I've compared with months of ARM VAP data and there are some slight differences due to different methods for gridding.  There are also some larger differences which upon investigation look like potential problems with the VAP.  These have been reported and are being looked at.  Currently there is a mean difference of -27m.

This method follows Heffter's criteria for finding PBL height with the same smoothing/gridding as was being performed in the Liu Liang method.  To this end, the overlapping preprocessing of the radiosonde data has been separated out into an individual function to simplify the codes.  This did cause some differences in the Liu Liang method, specifically the removal of the altitude smoothing.  The unit tests have been updated and the comparison with the VAP rerun yielding an average difference of 34m with a median of 5m.

image

### CODE REVIEW

File: act/retrievals/sonde.py [hunk 1]
```diff
@@ -477,3 +422,260 @@ def calculate_pbl_liu_liang(
...
+    top_inversion = []
+    for r in ranges:
+        if agl[r[1]] > 4000.:
+            continue
+        theta_diff = theta[r[1]] - theta[r[0]]
+        theta_diff_layer.append(theta_diff)
+        bottom_inversion.append(alt[r[0]])
+        top_inversion.append(alt[r[1]])
+        if pbl == 0 and theta_diff > 2.0:
+            # pbl = (alt[r[0]] + alt[r[1]]) / 2.
```
Comment (zssherman) [inline act/retrievals/sonde.py:524]: Can we get rid of this old code?

File: act/retrievals/sonde.py [hunk 2]
```diff
@@ -477,3 +422,260 @@ def calculate_pbl_liu_liang(
...
+            pbl = alt[r[0]]
+
+    if len(theta_diff_layer) == 0:
+        pbl = -9999.
+
+    # If PBL is not set, set it to the layer with the max theta diff
+    if pbl == 0:
+        idx = np.argmax(theta_diff_layer)
+        # pbl = (bottom_inversion[idx] + top_inversion[idx]) / 2.
```
Comment (zssherman) [inline act/retrievals/sonde.py:533]: Same here

File: act/retrievals/sonde.py [hunk 3]
```diff
@@ -477,3 +422,260 @@ def calculate_pbl_liu_liang(
...
+        group = list(map(itemgetter(1), group))
+        if group[-1] - group[0] > 0:
+            ranges.append((group[0], group[-1]))
+
+    # Subset ranges to lowest 5
+    if len(ranges) > 5:
+        ranges = ranges[0:5]
+
+    # For each layer, calculate the difference in theta from
+    # top and bottom of the layer.  The lowest layer where the
```
Comment (zssherman) [inline act/retrievals/sonde.py:510]: Multiple whitespace before the the, this is just more nitpicking though.

**Review by zssherman — CHANGES_REQUESTED** [review]
Just a few possible changes, some whitespace in the docstrings as well.

File: act/retrievals/sonde.py (hunk 3 above)
Comment (AdamTheisen) [reply act/retrievals/sonde.py:510]: I'm a double spacer between sentences! :)

**Comment by AdamTheisen** [comment]
This is strange.  These tests all run fine on my system but they seem to have different errors on different platforms

File: act/retrievals/sonde.py (hunk 3 above)
Comment (zssherman) [reply act/retrievals/sonde.py:510]: Ah then it's fine! I didn't know if that was intentional or accident! @AdamTheisen can resolve this then.

**Comment by zssherman** [comment]
@AdamTheisen That is odd, can you set pbl call to 0.0 instead of 0? Maybe some weird int to float behavior, highly doubt it, but worth a try

**Comment by zssherman** [comment]
@AdamTheisen Maybe with precision too, might need to you assert_close or something and set a tolerance for testing.

File: act/tests/test_retrievals.py [hunk 4]
```diff
@@ -228,6 +228,18 @@ def test_calculate_pbl_liu_liang():
         obj = act.retrievals.sonde.calculate_pbl_liu_liang(obj)
 
 
+def test_calculate_heffter_pbl():
+    files = glob.glob(act.tests.sample_files.EXAMPLE_TWP_SONDE_20060121)
```
Comment (mgrover1) [inline act/tests/test_retrievals.py:232]:
```suggestion
    files = sorted(glob.glob(act.tests.sample_files.EXAMPLE_TWP_SONDE_20060121))
```

File: act/tests/test_retrievals.py [hunk 5]
```diff
@@ -228,6 +228,18 @@ def test_calculate_pbl_liu_liang():
...
 
 
+def test_calculate_heffter_pbl():
+    files = sorted(glob.glob(act.tests.sample_files.EXAMPLE_TWP_SONDE_20060121))
+    ds = act.io.armfiles.read_netcdf(files[2])
+    ds['tdry'].attrs['units'] = 'degree_Celsius'
+    ds = act.retrievals.sonde.calculate_pbl_heffter(ds)
+    assert ds['pblht_heffter'].values == 960.
+    assert np.around(ds['atm_pres_ss'].values[1], 1) == 994.9
+    assert np.around(ds['potential_temperature_ss'].values[4], 1) == 298.4
```
Comment (mgrover1) [inline act/tests/test_retrievals.py:238]:
```suggestion
    assert np.testing.assert_almost_equal(ds['potential_temperature_ss'].values[4], 298.4, 1)
```

**Review by mgrover1 — CHANGES_REQUESTED** [review]
A couple of suggestions @AdamTheisen

File: act/tests/test_retrievals.py [hunk 6]
```diff
@@ -228,6 +228,18 @@ def test_calculate_pbl_liu_liang():
         obj = act.retrievals.sonde.calculate_pbl_liu_liang(obj)
 
 
+def test_calculate_heffter_pbl():
+    files = sorted(glob.glob(act.tests.sample_files.EXAMPLE_TWP_SONDE_20060121))
+    ds = act.io.armfiles.read_netcdf(files[2])
+    ds['tdry'].attrs['units'] = 'degree_Celsius'
+    ds = act.retrievals.sonde.calculate_pbl_heffter(ds)
+    assert ds['pblht_heffter'].values == 960.
+    assert np.testing.assert_almost_equal(ds['atm_pres_ss'].values[1], 994.9, 1)
```
Comment (mgrover1) [inline act/tests/test_retrievals.py:237]:
```suggestion
    np.testing.assert_almost_equal(ds['atm_pres_ss'].values[1], 994.9, 1)
```

Comment (mgrover1) [reply act/tests/test_retrievals.py:237]: This does not need an assert in front of it

File: act/tests/test_retrievals.py [hunk 7]
```diff
@@ -228,6 +228,18 @@ def test_calculate_pbl_liu_liang():
...
+    assert np.testing.assert_almost_equal(ds['potential_temperature_ss'].values[4], 298.4, 1)
```
Comment (mgrover1) [inline act/tests/test_retrievals.py:238]:
```suggestion
    np.testing.assert_almost_equal(ds['potential_temperature_ss'].values[4], 298.4, 1)
```

Comment (mgrover1) [reply act/tests/test_retrievals.py:238]: Same here!

**Review by mgrover1 — APPROVED** [review]
Great! Glad the CI issues are resolved.

**Review by zssherman — APPROVED** [review]
Looks good! @AdamTheisen Feel free to merge when ready.

**Pull request merged**